packages/engine-server/src/workspace/service.ts

```diff
@@ -849,36 +876,116 @@ export class WorkspaceService implements Disposable, IWorkspaceService {
+            try {
+              stashed = await git.stashCreate();
+              this.logger.info({ ctx, vaults, repo, stashed });
+              // this shouldn't fail, but for safety's sake
+              if (_.isEmpty(stashed) || !git.isValidStashCommit(stashed)) {
+                throw new DendronError({
+                  message: "unable to stash changes",
+                  payload: { stashed },
+                });
+              }
+              // stash create doesn't change the working directory, so we need to get rid of the tracked changes
+              await git.reset("hard");
+            } catch (err: any) {
+              this.logger.error({
+                ctx: "pullVaults",
+                vaults,
+                repo,
+                err,
+                stashed,
+              });
+              return makeResult(SyncActionStatus.CANT_STASH);
+            }
+          }
           try {
             await git.pull();
-            return { repo, vaults, status: SyncActionStatus.DONE };
+            if (stashed) {
+              const restored = await git.stashApplyCommit(stashed);
+              stashed = undefined;
+              if (!restored)
+                return makeResult(
+                  SyncActionStatus.MERGE_CONFLICT_AFTER_RESTORE
+                );
+            }
+            // pull went well, everything is in order. The finally block will restore any stashed changes.
+            return makeResult(SyncActionStatus.DONE);
           } catch (err: any) {
-            const stderr = err.stderr ? `: ${err.stderr}` : "";
-            throw new DendronError({
-              message: `error pulling vault${stderr}`,
-              payload: { err, repoPath: repo },
-            });
+            // Failed to pull, let's see why:
+            if (
+              (await git.hasMergeConflicts()) ||
+              (await git.hasRebaseInProgress())
+            ) {
+              if (stashed) {
+                // There was a merge conflict during the pull, and we have stashed changes.
+                // We can't apply the stash in this state, so we'd lose the users changes.
+                // Abort the rebase.
+                await git.rebaseAbort();
+                return makeResult(
+                  SyncActionStatus.MERGE_CONFLICT_LOSES_CHANGES
+                );
+              } else {
+                return makeResult(SyncActionStatus.MERGE_CONFLICT_AFTER_PULL);
+              }
+            } else {
+              const stderr = err?.stderr || "";
+              const vaultNames = vaults
+                .map((vault) => VaultUtils.getName(vault))
+                .join(",");
+              throw new DendronError({
+                message: `Failed to pull ${vaultNames}: ${stderr}`,
+                payload: {
+                  err,
+                  vaults,
+                  repo,
+                  stashed,
+                },
+              });
+            }
+          } finally {
```

Comment:
if the stash apply is successful, do we drop the stash? otherwise, would we end up with lots of stashed commits?

Reply:
`git stash create` doesn't actually create a stash entry like `git stash save` does. It only creates a dangling merge commit. If my understanding of `git gc` is correct, git will eventually automatically collect this commit since it's not referenced anywhere.